apps/rush-lib/src/logic/operations/OperationExecutionManager.ts

```diff
@@ -548,4 +588,131 @@ export class OperationExecutionManager {
     this._terminal.writeStdoutLine(leftPart + rightPart);
     this._terminal.writeStdoutLine('');
   }
+
+  /**
+   * Print a more detailed timeline and analysis of CPU usage for the build.
+   */
+  private _printTimeline(): void {
+    //
+    // Gather the operation records we'll be displaying
+    //
+
+    const operations: OperationExecutionRecord[] = [];
+
+    for (const operation of this._executionRecords) {
+      if (operation.stopwatch.startTime && operation.stopwatch.endTime) {
+        operations.push(operation);
+      }
+    }
+
+    operations.sort((a, b) => a.stopwatch.startTime! - b.stopwatch.startTime!);
+
+    //
+    // Determine timing for all tasks (wall clock and execution times)
+    //
+
+    const allStart: number = operations[0].stopwatch.startTime!;
+    const allEnd: number = Math.max(...operations.map((operation) => operation.stopwatch.endTime!));
+    const allDuration: number = allEnd - allStart;
+    const workDuration: number = operations
+      .map((operation) => operation.stopwatch.duration)
+      .reduce((sum, value) => sum + value);
+
+    //
+    // Do some calculations to determine what size timeline chart we need.
+    //
+
+    const maxNameLength: number = Math.max(...operations.map((operation) => operation.name.length));
+    const maxDurationLength: number = Math.max(
+      ...operations.map((operation) => operation.stopwatch.duration.toFixed(1).length + 1)
+    );
+    const chartWidth: number = TIMELINE_WIDTH - maxNameLength - maxDurationLength - 3;
+    //
+    // Loop through all operations, assembling some statistics about operations and
+    // phases, if applicable.
+    //
+
+    const output: string[] = [];
+
+    const durationByPhase: Map<string, number> = new Map();
+
+    const busyCpus: number[] = Array(this._parallelism).fill(-1);
+
+    output.push('='.repeat(TIMELINE_WIDTH));
+
+    for (const operation of operations) {
+      // Track time by phase
+      const phaseMatch: RegExpMatchArray | null = operation.name.match(/\((.+)\)$/);
```

Comment:
Feel free to forward the associated phase and associated project onto the OperationExecutionRecord and just read the value.